src/Compilers/VisualBasic/Portable/Syntax/SyntaxFactory.vb

```diff
@@ -532,5 +532,91 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             Return SyntaxEquivalence.AreEquivalent(oldList.Node, newList.Node, ignoreChildNode, topLevel:=False)
         End Function
 
+        ''' <summary>
+        ''' Determines if a syntax node is a part of a LINQ query.
+        ''' If so then a query is not terminated the syntax tree ends with an empty line.
+        ''' </summary>
+        ''' <returns>True if the token is part of a LINQ query and syntax tree ends with an empty line.</returns>
+        Private Shared Function IsPartOfUnterminatedLinqQuery(token As SyntaxToken, statementNode As SyntaxNode, endOfFileToken As SyntaxToken) As Boolean
+            Dim node = token.Parent
+            Do
+                If node.IsKind(SyntaxKind.QueryExpression) Then
+                    Return Not endOfFileToken.LeadingTrivia.Contains(Function(trivia) trivia.IsKind(SyntaxKind.EndOfLineTrivia))
+                End If
+
+                If node Is statementNode Then
+                    Exit Do
+                End If
+
+                node = node.Parent
+            Loop
+
+            Return False
+        End Function
+
+        ''' <summary>
+        ''' Determines if a submission is complete.
+        ''' Returns false if the syntax is valid but incomplete.
+        ''' Returns true if the syntax is invalid or complete.
+        ''' </summary>
+        ''' <param name="tree">Syntax tree.</param>
+        Public Shared Function IsCompleteSubmission(tree As SyntaxTree) As Boolean
+            Dim options As VisualBasicParseOptions = DirectCast(tree.Options, VisualBasicParseOptions)
+            Dim languageVersion As LanguageVersion = options.LanguageVersion
```

Comment:
Arguably, the question doesn't even make sense for regular code.  I think that leaves us with two options: either we throw an exception because the input is invalid or we return false because regular code is never a submission, complete or otherwise.